server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java

```diff
@@ -3228,4 +3232,109 @@ public static int parseIndexNameCounter(String indexName) {
             throw new IllegalArgumentException("unable to parse the index name [" + indexName + "] to extract the counter", e);
         }
     }
+
+    /**
+     * A subset of {@link IndexMetadata} storing only the shard count of an index
+     * Prior to v9.3, the entire {@link IndexMetadata} object was stored in heap and then loaded during snapshotting to determine
+     * the shard count. As per ES-12539, this is replaced with the {@link IndexShardCount} class that writes and loads only the index's
+     * shard count to and from heap memory, reducing the possibility of smaller nodes going OOMe during snapshotting
+     */
+    public static class IndexShardCount implements ToXContentFragment {
+        private static final String KEY_SHARD_COUNT = "shard_count";
+        private final int shardCount;
+
+        public IndexShardCount(int count) {
+            this.shardCount = count;
+        }
+
+        public int getCount() {
+            return shardCount;
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.field(KEY_SHARD_COUNT, shardCount);
+            return builder;
+        }
```

Comment:
I don't think we should have a `toXContent` here, we're only using this to read the shard count, not to write it out again